lib/fauna_model/base.rb

```diff
@@ -0,0 +1,269 @@
+module Fauna
+  module Model
+    class Base
+      include ActiveModel::Validations
+      include Fauna::Model::Dirty
+
+      class << self
+        def fauna_class
+          @fauna_class || fail(InvalidClass.new('No class set'))
+        end
+
+        def fauna_class=(fauna_class)
+          fauna_class = Fauna::Ref.new(fauna_class.to_s) unless fauna_class.is_a? Fauna::Ref
+          @fauna_class = fauna_class
+
+          field 'ref', 'ref', internal_readonly: true
+          field 'ts', 'ts', internal_readonly: true
+          field 'fauna_class', 'class', internal_readonly: true
+
+          case fauna_class
+          when 'databases'
+            field 'name', 'name'
+            field 'api_version', 'api_version'
+          when 'keys'
+            field 'database', 'database', internal_writeonce: true
+            field 'role', 'role', internal_writeonce: true
+            field 'secret', 'secret', internal_readonly: true
+            field 'hashed_secret', 'hashed_secret', internal_readonly: true
+          when 'classes'
+            field 'name', 'name'
+            field 'history_days', 'history_days'
+            field 'ttl_days', 'ttl_days'
+            field 'permissions', 'permissions'
+          when 'indexes'
+            field 'name', 'name'
+            field 'source', 'source'
+            field 'terms', 'terms'
+            field 'values', 'values'
+            field 'unique', 'unique'
+            field 'permissions', 'permissions'
+          end
+        end
+
+        def fields
+          @fields ||= {}
```

Comment:
You could create a class initializer (via `def inherited` I think?) to do this sort of initialization eagerly. Then just use attr_reader.